### PR TITLE
TopNavBar component

### PR DESCRIPTION
**Pull request by ke1k0**

### CODE REVIEW

File: packages/ui-top-nav-bar/src/TopNavBar/TopNavBarLayout/props.ts [hunk 1]
```diff
@@ -49,83 +47,163 @@ import type { BrandChild } from '../TopNavBarBrand/props'
...
+}
+
+type TopNavBarLayoutOwnProps = CommonTopNavBarLayoutProps & {
+  /**
+   * Config object for the "desktop" mode:
+   */
+  desktopConfig?: DesktopLayoutOwnProps
+
+  /**
+   * Config object for the "small viewport" mode. It is required, if there is no custom `renderSmallViewportLayout` is set on `<TopNavBar>`.
```
Comment (Brailor) [inline packages/ui-top-nav-bar/src/TopNavBar/TopNavBarLayout/props.ts:113]: There is no `renderSmallViewportLayout` prop on the `TopNavBar` component, and there seems to be no related code in the package, maybe this is an outdated comment or the functionality is missing?

Comment (ke1k0) [reply packages/ui-top-nav-bar/src/TopNavBar/TopNavBarLayout/props.ts]: Thanks, it was outdated, fixed!

File: packages/shared-types/src/CommonTypes.ts [hunk 2]
```diff
@@ -53,6 +56,17 @@ export type Renderable<P = never> =
...
+/**
+ * Union type helper for the "children" prop.
+ * The argument has to be type of, or union of type of
+ * React.ComponentElement<any, any> */
+export type ChildrenOfType<T extends ComponentElement<any, any>> =
+  | T
+  | undefined
+  | null
+  | (T | undefined | null)[]
+  | (T | undefined | null | (T | undefined | null)[])[]
```
Comment (Brailor) [inline packages/shared-types/src/CommonTypes.ts:68]: If we don't care about restricting the depth of nesting of arrays, then we can rewrite this type as:

```ts
type ChildrenOfType<T extends ComponentElement<any, any>> =
| T 
| undefined
| null
| ChildrenOfType<T>[]
```

[Example](https://www.typescriptlang.org/play?#code/C4TwDgpgBAwg9gWzHAdhFwCiAbCD3AA8A0lBAB7DoAmAzlLcAE4CWKA5gDRQBqUAvFACGKEAD4BUAEoQAxnCbUSnHmICwAKAD0AKh2aoOqAFUULVFFCQoACwjZITKADMFlu1ABEsmy2zUmdE8oMCY4MAA6AyMAFQ8hJnYAV3wMWyF6YDgoACNoK2g4Z243JLMLIstwQudo6QghWWAI+CRUAhw8AkIREG5eiR0tTQpkJmAq6xhff0CUAHlnGOrCGLJKGnpW5DQMTtSiXv7RMQl+AygAHygYi+uy6ghnNghqO6gUJOxsd4AKNfuKEezzQ1CuHy+2AAlABtAC6fwBUAeTxeYOun2+4P+4JRINe4Mx0PhsIRGk0BVgMwC6EWy0gACZVusqECtogdh1cAceqJjuIzpprmshcigajQaKiaLpn4aQsltUmTExPDNOqNPIUIwoE0AFxUuVzOkrADenjs3zgngNnmAdkCngAvmcoDDTZbsHBbfaII6ndx3Z7vV5ff7uB77F6fQ6IM7uES4YGg1GQ3bY864VmNVqdU0GQbZbNaYrGYRzcGbaGMy7JCmrTG-XGA27Iw3q0341A29GO+GId8k26idw8WjOPXe+nO06s2SgA)

Comment (ke1k0) [reply packages/shared-types/src/CommonTypes.ts:68]: It seems that React doesn't care about the depth, so I'll update it! Thanks, good tip!

**Review by Brailor — APPROVED** [review]
Really solid work!

**Review by HerrTopi — APPROVED** [review]

**Pull request merged**